File: articles/dark-traits-virtue-and-the-moral-structure-of-personality/python/analyze_dark_traits_virtue.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_dark_traits_virtue_personality.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

DARK = ["machiavellianism", "narcissism", "psychopathy", "sadism"]
VIRTUE = ["honesty_humility", "conscientious_reliability", "fairness_orientation", "compassion_kindness"]
LAYERS = ["moral_identity", "practical_judgment", "institutional_accountability", "status_reward_pressure"]
OUTCOMES = ["unethical_behavior", "prosocial_restraint", "harm_indicator"]

def ols(df, y, xs):
    X = np.column_stack([np.ones(len(df)), df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, df[y].to_numpy(float), rcond=None)
    return dict(zip(["intercept", *xs], map(float, beta)))

df = pd.read_csv(DATA)
required = {"participant_id", "institutional_context", *DARK, *VIRTUE, *LAYERS, *OUTCOMES}
missing = required - set(df.columns)
if missing:
    raise ValueError(f"Missing columns: {sorted(missing)}")

df["dark_trait_burden"] = df[DARK].mean(axis=1)
df["virtue_relevant_tendency"] = df[VIRTUE].mean(axis=1)
df["moral_integration_index"] = (df["virtue_relevant_tendency"] + df["moral_identity"] + df["practical_judgment"]) / 3
df["dark_accountability_risk"] = df["dark_trait_burden"] * (7 - df["institutional_accountability"])
df["dominant_dark_trait"] = df[DARK].idxmax(axis=1)
df["dark_level"] = np.where(df["dark_trait_burden"] > df["dark_trait_burden"].median(), "higher_darkness", "lower_darkness")
df["virtue_level"] = np.where(df["virtue_relevant_tendency"] > df["virtue_relevant_tendency"].median(), "higher_virtue_relevant", "lower_virtue_relevant")
df["moral_profile"] = df["dark_level"] + "_" + df["virtue_level"]

df.groupby("institutional_context").agg(
    n=("participant_id", "count"),
    dark_trait_burden_mean=("dark_trait_burden", "mean"),
    virtue_relevant_mean=("virtue_relevant_tendency", "mean"),
    accountability_mean=("institutional_accountability", "mean"),
    status_pressure_mean=("status_reward_pressure", "mean"),
    unethical_behavior_mean=("unethical_behavior", "mean"),
    prosocial_restraint_mean=("prosocial_restraint", "mean"),
    harm_indicator_mean=("harm_indicator", "mean"),
).reset_index().to_csv(OUT / "python_institutional_context_summary.csv", index=False)

df.groupby("moral_profile").agg(
    n=("participant_id", "count"),
    dark_trait_burden_mean=("dark_trait_burden", "mean"),
    virtue_relevant_mean=("virtue_relevant_tendency", "mean"),
    moral_integration_mean=("moral_integration_index", "mean"),
    unethical_behavior_mean=("unethical_behavior", "mean"),
    prosocial_restraint_mean=("prosocial_restraint", "mean"),
    harm_indicator_mean=("harm_indicator", "mean"),
).reset_index().to_csv(OUT / "python_moral_profile_summary.csv", index=False)

df.groupby("dominant_dark_trait").agg(
    n=("participant_id", "count"),
    dark_trait_burden_mean=("dark_trait_burden", "mean"),
    unethical_behavior_mean=("unethical_behavior", "mean"),
    harm_indicator_mean=("harm_indicator", "mean"),
    virtue_relevant_mean=("virtue_relevant_tendency", "mean"),
).reset_index().to_csv(OUT / "python_dominant_dark_trait_summary.csv", index=False)

corr_cols = DARK + VIRTUE + LAYERS + OUTCOMES + ["dark_trait_burden", "virtue_relevant_tendency", "moral_integration_index", "dark_accountability_risk"]
df[corr_cols].corr().to_csv(OUT / "python_dark_traits_virtue_correlations.csv")

models = {
    "unethical_behavior": ols(df, "unethical_behavior", DARK + VIRTUE),
    "prosocial_restraint": ols(df, "prosocial_restraint", ["dark_trait_burden", "virtue_relevant_tendency", "moral_identity", "practical_judgment"]),
    "harm_indicator": ols(df, "harm_indicator", ["dark_trait_burden", "virtue_relevant_tendency", "institutional_accountability", "status_reward_pressure", "dark_accountability_risk"]),
    "note": "Synthetic demonstration only; not moral ranking, hiring guidance, legal evaluation, or social sorting."
}
(OUT / "python_dark_traits_virtue_models.json").write_text(json.dumps(models, indent=2))
df.to_csv(OUT / "python_scored_dark_traits_virtue.csv", index=False)
print(f"Wrote outputs to {OUT}")

File: articles/dark-traits-virtue-and-the-moral-structure-of-personality/r/analyze_dark_traits_virtue.R
root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_dark_traits_virtue_personality.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)
dark <- c("machiavellianism", "narcissism", "psychopathy", "sadism")
virtue <- c("honesty_humility", "conscientious_reliability", "fairness_orientation", "compassion_kindness")

data$dark_trait_burden <- rowMeans(data[, dark])
data$virtue_relevant_tendency <- rowMeans(data[, virtue])
data$moral_integration_index <- rowMeans(data[, c("virtue_relevant_tendency", "moral_identity", "practical_judgment")])
data$dark_accountability_risk <- data$dark_trait_burden * (7 - data$institutional_accountability)
data$dominant_dark_trait <- dark[max.col(data[, dark], ties.method = "first")]
data$moral_profile <- paste(
  ifelse(data$dark_trait_burden > median(data$dark_trait_burden), "higher_darkness", "lower_darkness"),
  ifelse(data$virtue_relevant_tendency > median(data$virtue_relevant_tendency), "higher_virtue_relevant", "lower_virtue_relevant"),
  sep = "_"
)

context_summary <- aggregate(
  data[, c("dark_trait_burden", "virtue_relevant_tendency", "institutional_accountability", "status_reward_pressure", "unethical_behavior", "prosocial_restraint", "harm_indicator")],
  by = list(institutional_context = data$institutional_context),
  FUN = mean
)
write.csv(context_summary, file.path(outputs, "r_institutional_context_summary.csv"), row.names = FALSE)

profile_summary <- aggregate(
  data[, c("dark_trait_burden", "virtue_relevant_tendency", "moral_integration_index", "unethical_behavior", "prosocial_restraint", "harm_indicator")],
  by = list(moral_profile = data$moral_profile),
  FUN = mean
)
write.csv(profile_summary, file.path(outputs, "r_moral_profile_summary.csv"), row.names = FALSE)

corr_cols <- c(dark, virtue, "moral_identity", "practical_judgment", "institutional_accountability", "status_reward_pressure", "unethical_behavior", "prosocial_restraint", "harm_indicator", "dark_trait_burden", "virtue_relevant_tendency", "moral_integration_index", "dark_accountability_risk")
write.csv(cor(data[, corr_cols], use = "pairwise.complete.obs"), file.path(outputs, "r_dark_traits_virtue_correlations.csv"))

model_unethical <- lm(unethical_behavior ~ machiavellianism + narcissism + psychopathy + sadism + honesty_humility + conscientious_reliability + fairness_orientation + compassion_kindness, data = data)
model_prosocial <- lm(prosocial_restraint ~ dark_trait_burden + virtue_relevant_tendency + moral_identity + practical_judgment, data = data)
model_harm <- lm(harm_indicator ~ dark_trait_burden + virtue_relevant_tendency + institutional_accountability + status_reward_pressure + dark_accountability_risk, data = data)

capture.output(summary(model_unethical), file = file.path(outputs, "r_unethical_behavior_model_summary.txt"))
capture.output(summary(model_prosocial), file = file.path(outputs, "r_prosocial_restraint_model_summary.txt"))
capture.output(summary(model_harm), file = file.path(outputs, "r_harm_indicator_model_summary.txt"))

write.csv(data, file.path(outputs, "r_scored_dark_traits_virtue.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")
